File: elixir/space-age/space_age.exs
defmodule SpaceAge do
  @type planet :: :mercury | :venus | :earth | :mars | :jupiter
                | :saturn | :uranus | :neptune

  @spec in_seconds(float) :: float
  defp in_seconds(year) when is_float(year), do: year * 365.25 * 24 * 60 * 60

  @spec orbital_period_in_seconds(planet) :: float
  defp orbital_period_in_seconds(planet) do
    case planet do
      :earth -> 1.0 |> in_seconds
      :mercury -> 0.2408467 |> in_seconds
      :venus -> 0.61519726 |> in_seconds
      :mars -> 1.8808158 |> in_seconds
      :jupiter -> 11.862615 |> in_seconds
      :saturn -> 29.447498 |> in_seconds
      :uranus -> 84.016846 |> in_seconds
      :neptune -> 164.79132 |> in_seconds
      _ -> raise "Unknown planet :#{planet}"
    end
  end
    
  @doc """
  Return the number of years a person that has lived for 'seconds' seconds is
  aged on 'planet'.
  """
  @spec age_on(planet, pos_integer) :: float
  def age_on(planet, seconds), do: seconds / orbital_period_in_seconds(planet)
end

File: teams/microsoft-cse-golang-study-group-summer-2019/go/space-age/space_age.go
package space

func init() {
	orbitalPeriodOf = createOrbitalTable()
}

func createOrbitalTable() map[Planet]float64 {
	planetYearInSeconds := func(orbitalPeriodDurationInEarthYears float64) float64 {
		const (
			earthYearInSeconds = 60 * 60 * 24 * 365.25
		)
		return earthYearInSeconds * orbitalPeriodDurationInEarthYears
	}

	return map[Planet]float64{
		"Earth":   planetYearInSeconds(1.0),
		"Mercury": planetYearInSeconds(0.2408467),
		"Venus":   planetYearInSeconds(0.61519726),
		"Mars":    planetYearInSeconds(1.8808158),
		"Jupiter": planetYearInSeconds(11.862615),
		"Saturn":  planetYearInSeconds(29.447498),
		"Uranus":  planetYearInSeconds(84.016846),
		"Neptune": planetYearInSeconds(164.79132),
	}
}

var orbitalPeriodOf map[Planet]float64

// Planet represents a planet
type Planet string

// Age is the age of a planet.
func Age(seconds float64, planet Planet) float64 {
	return seconds / orbitalPeriodOf[planet]
}
